src/trio/testing/_raises_group.py

```diff
@@ -0,0 +1,305 @@
+from __future__ import annotations
+
+import re
+import sys
+from typing import (
+    TYPE_CHECKING,
+    Callable,
+    ContextManager,
+    Generic,
+    Iterable,
+    Pattern,
+    TypeVar,
+    cast,
+)
+
+from trio._util import final
+
+if TYPE_CHECKING:
+    # sphinx will *only* work if we use types.TracebackType, and import
+    # *inside* TYPE_CHECKING. No other combination works.....
+    import types
+
+    from typing_extensions import TypeGuard
+
+if sys.version_info < (3, 11):
+    from exceptiongroup import BaseExceptionGroup
+
+E = TypeVar("E", bound=BaseException)
+
+
+# minimal version of pytest.ExceptionInfo in case it is not available
+@final
+class _ExceptionInfo(Generic[E]):
+    _excinfo: tuple[type[E], E, types.TracebackType] | None
+
+    def __init__(self, excinfo: tuple[type[E], E, types.TracebackType] | None):
+        self._excinfo = excinfo
+
+    def fill_unfilled(self, exc_info: tuple[type[E], E, types.TracebackType]) -> None:
+        """Fill an unfilled ExceptionInfo created with ``for_later()``."""
+        assert self._excinfo is None, "ExceptionInfo was already filled"
+        self._excinfo = exc_info
+
+    @classmethod
+    def for_later(cls) -> _ExceptionInfo[E]:
+        """Return an unfilled ExceptionInfo."""
+        return cls(None)
+
+    @property
+    def type(self) -> type[E]:
+        """The exception class."""
+        assert (
+            self._excinfo is not None
+        ), ".type can only be used after the context manager exits"
+        return self._excinfo[0]
+
+    @property
+    def value(self) -> E:
+        """The exception value."""
+        assert (
+            self._excinfo is not None
+        ), ".value can only be used after the context manager exits"
+        return self._excinfo[1]
+
+    @property
+    def tb(self) -> types.TracebackType:
+        """The exception raw traceback."""
+        assert (
+            self._excinfo is not None
+        ), ".tb can only be used after the context manager exits"
+        return self._excinfo[2]
+
+
+# this may bite users with type checkers not using pytest, but that should
+# be rare and give quite obvious errors in tests trying to do so.
+# Ignoring "incorrect import of pytest", it makes sense in this situation.
+if TYPE_CHECKING:
```

Comment:
If the class above is sufficient, why use the pytest version at all?

Reply:
[pytest.ExceptionInfo](https://docs.pytest.org/en/stable/reference/reference.html#exceptioninfo) has a couple helper methods I didn't bother reproducing - `match`, `getrepr`, `exconly`. And somebody writing tests might very well want to make use of those after catching an `ExceptionGroup`

Reply:
realized that anybody using this without pytest would have zero types on exceptioninfo, so reversed it to always pretend we return `_ExceptionInfo` and added stubs for the helper methods. Maybe easier to simply always use `_ExceptionInfo`.